model/change_point_detection.go

```diff
@@ -328,6 +341,94 @@ func ReplaceChangePoints(ctx context.Context, env cedar.Environment, performance
 	return catcher.Resolve()
 }
 
+func GetTotalPagesForChangePointsGroupedByVersion(ctx context.Context, env cedar.Environment, projectId string, pageSize int) (int, error) {
+	type CountResult struct {
+		Count int `bson:"count"`
+	}
+	countKey := "count"
+	pipe := []bson.M{
+		{
+			"$count": countKey,
+		},
+	}
+	pipe = append(getMatchAndUnwindForChangePointsGroupedByVersion(projectId), pipe...)
+	cur, err := env.GetDB().Collection(perfResultCollection).Aggregate(ctx, pipe)
+	if err != nil {
+		return 0, errors.Wrap(err, "Cannot aggregate to get count of change points grouped by version")
+	}
+	defer cur.Close(ctx)
+	var res CountResult
+	if cur.Next(ctx) {
+		err = cur.Decode(&res)
+		if err != nil {
+			return 0, errors.Wrap(err, "Cannot decode response of getting count of change points grouped by version")
+		}
+		return int(math.Ceil(float64(res.Count) / float64(pageSize))), nil
+	}
+	return 0, errors.New("Not able to get count of total changepoints matching query")
+}
+
+func getMatchAndUnwindForChangePointsGroupedByVersion(projectId string) []bson.M {
+	return []bson.M{
+		{
+			"$match": bson.M{
+				bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoProjectKey):             projectId,
+				bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, "0"): bson.M{"$exists": true},
+			},
+		},
+		{
+			"$unwind": "$" + bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey),
+		},
+	}
+}
+
+func GetChangePointsGroupedByVersion(ctx context.Context, env cedar.Environment, projectId string, page, pageSize int) ([]GetChangePointsGroupedByVersionResult, error) {
+	pipe := []bson.M{
+		{
+			"$sort": bson.M{
+				bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, perfChangePointCalculatedOnKey): -1,
+			},
+		},
+		{
+			"$skip": page * pageSize,
+		},
+		{
+			"$limit": pageSize,
+		},
+		{
+			"$project": bson.M{
+				"perf_result_id": "$_id",
+				"change_point":   "$" + bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey),
+				"project":        "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoProjectKey),
+				"variant":        "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoVariantKey),
+				"task":           "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoTaskNameKey),
+				"test":           "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoTestNameKey),
+				"thread_level":   "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfResultInfoArgumentsKey, "thread_level"),
+				"version":        "$" + bsonutil.GetDottedKeyName(perfInfoKey, perfVersionlKey),
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id":           "$version",
+				"change_points": bson.M{"$push": "$$ROOT"},
+				"version_id":    bson.M{"$first": "$version"},
+			},
+		},
+	}
+	pipe = append(getMatchAndUnwindForChangePointsGroupedByVersion(projectId), pipe...)
```

Comment:
can you have the builder functions be variadic bson.M functions that return []bson.M so the ordering is a bit more clear?